lib/rubocop/cop/rspec/stubbed_mock.rb

```diff
@@ -0,0 +1,97 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks that message expectations do not have a configured response.
+      #
+      # @example
+      #
+      #   # bad
+      #   expect(foo).to receive(:bar).with(42).and_return("hello world")
+      #
+      #   # good (without spies)
+      #   allow(foo).to receive(:bar).with(42).and_return("hello world")
+      #   expect(foo).to receive(:bar).with(42)
+      #
+      class StubbedMock < Base
+        MSG = 'Prefer `%<replacement>s` to `%<method_name>s` when ' \
+              'configuring a response.'
+
+        def_node_matcher :message_expectation?, <<-PATTERN
+          {
+            (send nil? { :receive :receive_message_chain } ...)
+            (send (send nil? :receive ...) :with ...)
+          }
+        PATTERN
+
+        def_node_matcher :configured_response?, <<~PATTERN
+          { :and_return :and_raise :and_throw :and_yield
+            :and_call_original :and_wrap_original }
+        PATTERN
+
+        def self.expectation_with(matcher)
```

Comment:
Nice :1st_place_medal: 
Up to now I'd have a matcher `expectation` that caputres `matcher` and then some code like
```
expectation(node) do |matcher|
  next unless configured_response?(matcher)
...
end
```
Now there's one more tool in my toolbox

Reply:
Now when I think of it, I like your variant more.

```diff
-        def_node_matcher :expectation_with_configured_response,
-                         expectation_with(<<~PATTERN)
-                           (send #message_expectation? #configured_response? _)
-                         PATTERN
+        def_node_matcher :matcher_with_configured_response, <<~PATTERN
+          (send #message_expectation? #configured_response? _)
+        PATTERN
```
and the following becomes slightly more generic:
```
def on_send(node)
  expectation(node, &:on_expectation) # this can become a concern!
end

def on_expectation(expectation)
  matcher_with_configured_response(expectation) do |matcher, method_name|
    add_offense(matcher, message: msg(method_name))
  end

  matcher_with_return_block(expectation) do |matcher, method_name|
    add_offense(matcher, message: msg(method_name))
  end

  matcher_with_blockpass_or_hash(expectation) do |matcher, method_name|
    add_offense(matcher, message: msg(method_name))
  end
end
```

I'll give it a shot later today.

Reply:
Worked! I like it much better this way.

Thanks, and I think there are cases when my old approach would work, but in this specific cop `on_expectation` worked quite nice.

I also had this crazy idea https://github.com/rubocop-hq/rubocop-ast/issues/97, but it ended up being even more pleasant to look at.